### PR TITLE
[Clang] Allow the value of unroll count to be zero  in `#pragma GCC unroll` and `#pragma unroll`

### LINKED ISSUE #88624: `#pragma GCC unroll 0` not allowed: `invalid value '0'; must be positive`
`#pragma GCC unroll 0` does not compile, as apparently the specified number needs to be positive.

```
/app/main.cpp:4:24: error: invalid value '0'; must be positive
    4 |     #pragma GCC unroll 0
      |                        ^
1 error generated.
```

However, from <https://gcc.gnu.org/onlinedocs/gcc/Loop-Specific-Pragmas.html>:
> The values of 0 and 1 block any unrolling of the loop.

The program also successfully compiles with GCC, only `clang` seems to have this implemented incorrectly.

Godbolt: https://godbolt.org/z/bEPo64GsP

### PR DESCRIPTION
**Pull request by yronglin**
Fixes https://github.com/llvm/llvm-project/issues/88624

GCC allows the value of loop unroll count to be zero, and the values of 0 and 1 block any unrolling of the loop. This PR aims to make clang keeps the same behavior with GCC.
https://gcc.gnu.org/onlinedocs/gcc/Loop-Specific-Pragmas.html

### CODE REVIEW

**Comment by Endilll** [comment]
`Sema.h` changes look good.

File: clang/docs/ReleaseNotes.rst
Comment (tbaederr) [inline clang/docs/ReleaseNotes.rst:426]:
```suggestion
- Clang now allows the value of unroll count to be zero in ``#pragma GCC unroll`` and ``#pragma unroll``.
```

Comment (yronglin) [reply clang/docs/ReleaseNotes.rst]: Done!

File: clang/lib/Sema/SemaExpr.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaExpr.cpp:3912]: I'd very much like to keep this checking of the name OUT of Sema here, and have the function just take a bool for `IsPragmaUnroll` or something.

**Review by erichkeane — COMMENTED** [review]
1 nit, else this lgtm.

File: clang/lib/Sema/SemaExpr.cpp
Comment (yronglin) [reply clang/lib/Sema/SemaExpr.cpp]: Thanks for your review! I've add a bool parameter `AllowZero` here. I want to go further, should we move all the string checks in `Sema::handleLoopHintAttr` to `Parser::HandlePragmaLoopHint`? (https://github.com/llvm/llvm-project/blob/ac791888bbbe58651e597cf7a4b2276424b77a92/clang/lib/Sema/SemaStmtAttr.cpp#L108) , these checks used to classify `LoopHintAttr::OptionType` and `LoopHintAttr::LoopHintState`, I'd like to classification in `Parser::HandlePragmaLoopHint` and pass the result into `Sema::handleLoopHintAttr` through `ParsedAttr`.

Comment (erichkeane) [reply clang/lib/Sema/SemaExpr.cpp]: Thanks!  I don't think we should do those changes in THIS patch, but I'd be ok with a followup patch to do that movement.

**Review by erichkeane — APPROVED** [review]

File: clang/lib/Sema/SemaExpr.cpp
Comment (yronglin) [reply clang/lib/Sema/SemaExpr.cpp]: Thanks, I'll do it in a separate patch!

**Comment by yronglin** [comment]
Wait for CI green, but seems windows bots fall into a trouble.

**Comment by erichkeane** [comment]
> Wait for CI green, but seems windows bots fall into a trouble.

Windows bot is typically pretty far behind(usually 4-5 hrs during the day), but 13 hours is a heck of a delay!  I think all we can do is wait :/

**Review by AaronBallman — APPROVED** [review]
LGTM assuming precommit CI on Windows eventually comes back green.

**Comment by yronglin** [comment]
> LGTM assuming precommit CI on Windows eventually comes back green.

Thanks for your review!

**Comment by yronglin** [comment]
Hope CI can be come back soon

**Comment by yronglin** [comment]
Thanks for all the reviewers! The last CI was green, but there was a conflict in ReleaseNotes. After I merged the main branch and resolved the conflict, there was a build issue with CI, but it was not caused by this PR(Seems the root cause is flang). therefor, I'll merge this PR.

**Comment by alexfh** [comment]
After this patch clang doesn't accept integer template argument as a parameter of `#pragma unroll` (https://gcc.godbolt.org/z/Woc7zs3sK):
```
template<int Unroll>
void test(int *List, int Length) {
  int i = 0;
#pragma unroll Unroll
  while (i + 1 < Length) {
    List[i] = i;
  }
}
```

```
<source>:4:16: error: expression is not an integral constant expression
    4 | #pragma unroll Unroll
      |                ^~~~~~
```

Please fix soon or revert. Thanks!

**Comment by ronlieb** [comment]
our downstream CI for amdgpu build of rccl, rocblas and rocSolver are seeing this assertions due to this patch
[2024-04-21T03:28:05.859Z] clang-19: /jenkins/workspace/compiler-psdb-amd-staging/repos/external/llvm-project/clang/lib/AST/ExprConstant.cpp:15721: bool clang::Expr::EvaluateAsRValue(clang::Expr::EvalResult&, const clang::ASTContext&, bool) const: Assertion `!isValueDependent() && "Expression evaluator can't be called on a dependent expression."' failed.

if a reproducer is needed, we can try and get one , bit of effort. let me know.

**Comment by Endilll** [comment]
@ronlieb nice and small reproducer would definitely help resolving this.

**Comment by shiltian** [comment]
@alexfh @ronlieb @Endilll fix in https://github.com/llvm/llvm-project/pull/89567.

**Comment by yronglin** [comment]
Thank you report this issue! @ronlieb Could you please provide a simple reproducer, we can use it to strengthen clang's test.

**Comment by alexfh** [comment]
Thanks for the prompt fix!

**Comment by alexfh** [comment]
There's one more questionable effect of this commit (together with the fix in #89567). Some of our code started triggering "loop not unrolled" warnings. A reduced example is at https://godbolt.org/z/jG4xsGY7s
```
#include <cstddef>
#include <bit>

template <bool Flag>
int FailToBuild(int n) {
  constexpr int N = 100;
  auto init = [=]() { return Flag ? n : 0UL; };
  auto cond = [=](size_t ix) { return Flag ? ix != 0 : ix < 10; };
  auto iter = [=](size_t ix) {
    return Flag ? ix & ~(1ULL << std::countr_zero(ix)) : ix + 1;
  };
#pragma unroll Flag ? 1 : N
  for (size_t ix = init(); cond(ix); ix = iter(ix)) {
    n *= n;
  }
  return n;
}

int foo(int n) {
    return FailToBuild<true>(n);
}

int bar(int n) {
    return FailToBuild<false>(n);
}
```

`clang -std=c++20 -O3` started generating the following warning on this code after the commit:
```
<source>:13:3: warning: loop not unrolled: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Wpass-failed=transform-warning]
   13 |   for (size_t ix = init(); cond(ix); ix = iter(ix)) {
```

It doesn't seem the generated code has actually changed, thus, the question is whether these are legit warnings that used to be missed, or vice versa?

**Comment by yronglin** [comment]
> There's one more questionable effect of this commit (together with the fix in #89567). Some of our code started triggering "loop not unrolled" warnings. A reduced example is at https://godbolt.org/z/jG4xsGY7s
> 
> ```
> #include <cstddef>
> #include <bit>
> 
> template <bool Flag>
> int FailToBuild(int n) {
>   constexpr int N = 100;
>   auto init = [=]() { return Flag ? n : 0UL; };
>   auto cond = [=](size_t ix) { return Flag ? ix != 0 : ix < 10; };
>   auto iter = [=](size_t ix) {
>     return Flag ? ix & ~(1ULL << std::countr_zero(ix)) : ix + 1;
>   };
> #pragma unroll Flag ? 1 : N
>   for (size_t ix = init(); cond(ix); ix = iter(ix)) {
>     n *= n;
>   }
>   return n;
> }
> 
> int foo(int n) {
>     return FailToBuild<true>(n);
> }
> 
> int bar(int n) {
>     return FailToBuild<false>(n);
> }
> ```
> 
> `clang -std=c++20 -O3` started generating the following warning on this code after the commit:
> 
> ```
> <source>:13:3: warning: loop not unrolled: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Wpass-failed=transform-warning]
>    13 |   for (size_t ix = init(); cond(ix); ix = iter(ix)) {
> ```
> 
> It doesn't seem the generated code has actually changed, thus, the question is whether these are legit warnings that used to be missed, or vice versa?



> There's one more questionable effect of this commit (together with the fix in #89567). Some of our code started triggering "loop not unrolled" warnings. A reduced example is at https://godbolt.org/z/jG4xsGY7s
> 
> ```
> #include <cstddef>
> #include <bit>
> 
> template <bool Flag>
> int FailToBuild(int n) {
>   constexpr int N = 100;
>   auto init = [=]() { return Flag ? n : 0UL; };
>   auto cond = [=](size_t ix) { return Flag ? ix != 0 : ix < 10; };
>   auto iter = [=](size_t ix) {
>     return Flag ? ix & ~(1ULL << std::countr_zero(ix)) : ix + 1;
>   };
> #pragma unroll Flag ? 1 : N
>   for (size_t ix = init(); cond(ix); ix = iter(ix)) {
>     n *= n;
>   }
>   return n;
> }
> 
> int foo(int n) {
>     return FailToBuild<true>(n);
> }
> 
> int bar(int n) {
>     return FailToBuild<false>(n);
> }
> ```
> 
> `clang -std=c++20 -O3` started generating the following warning on this code after the commit:
> 
> ```
> <source>:13:3: warning: loop not unrolled: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Wpass-failed=transform-warning]
>    13 |   for (size_t ix = init(); cond(ix); ix = iter(ix)) {
> ```
> 
> It doesn't seem the generated code has actually changed, thus, the question is whether these are legit warnings that used to be missed, or vice versa?

Thank you for your report, I will find out the cause as soon as possible.

**Comment by yronglin** [comment]
Sorry, I didn't find this problem last time I reviewed https://github.com/llvm/llvm-project/pull/89567. @alexfh Seems this issue introduced by https://github.com/llvm/llvm-project/pull/89567, I've a PR to fix it https://github.com/llvm/llvm-project/pull/90240.

**Comment by yronglin** [comment]
@alexfh fixed.

**Pull request merged**
